### DISCUSSION TITLE
code to prevent label switching among imputation fits

### DISCUSSION THREAD

**Discussion by AnnaBorroni** [bmcclintock/momentuHMM#184] [Q&A] [answered]
Hi everyone!

Examining my individual model fits,  I found that labels switched among imputations. To avoid this problem I have seen that I can use DM, workbounds and userBounds but I am not sure if I completely understood how they work. I would like to force the step length of ARS to be smaller than Transit's step length and both being shorter than 10 km and the angle concentration value (Von Mises distribution) of ARS being smaller than Transit' angle concentration value. Is the following script correct?

nbStates <- 2

stateNames <- c("ARS","Transit")

dist = list(step = "gamma", angle = "vm")

stepDM <- matrix(c(1,0,0,0,
                   1,1,0,0,
                   0,0,1,0,
                   0,0,0,1),
                 nrow = 2*nbStates,
                 byrow = T,
                 dimnames = list(c(paste0("mu_", 1:nbStates),
                                   paste0("sd_", 1:nbStates)),
                                 c("wmu_1","wmu_2",
                                   paste0("wsd_", 1:nbStates))))

angleDM <- matrix(c(1,0,
                    1,1),
                  nrow=nbStates,
                  byrow=TRUE,
                  dimnames=list(paste0("concentration_",1:nbStates),
                                paste0("wconc_", 1:nbStates)))

stepworkBounds <- matrix(c(-Inf,Inf,
                              0,Inf,
                           -Inf,Inf,
                           -Inf,Inf),nrow=ncol(stepDM),byrow = TRUE,dimnames=list(c("wmu_12","wmu_2",
                                                                                    paste0("wsd_", 1:nbStates)),
                                                                                  c("lower","upper")))

angleworkBounds <- matrix(c(-Inf,Inf,
                               0,Inf),nrow=ncol(angleDM),byrow = TRUE,dimnames=list(c(paste0("wconc_", 1:nbStates)),
                                                                                  c("lower","upper")))

stepBounds <- matrix(c(0,10,
                     0,10,
                     0,10,
                     0,10),nrow=2*nbStates,byrow=TRUE, dimnames=list(c(paste0("mu_", 1:nbStates),
                                                                          paste0("sd_", 1:nbStates)),c("lower","upper")))

angleBounds<-matrix(c(0,Inf,
                      0,Inf),nrow=nbStates,byrow=TRUE, dimnames=list(rownames(angleDM),c("lower","upper")))

DM <- list (step = stepDM, angle = angleDM)
workBounds <- list(step = stepworkBounds, angle = angleworkBounds)
userBounds <- list(step = stepBounds, angle = angleBounds)

Thank you very much in advance!

**Comment by bmcclintock** [comment, marked as answer]
Apologies for the delay, I've been in the field and on vacation for much of the past three months. The best way to confirm the code is achieving what you want is to write out the actual equations and verify the fitted model output is indeed enforcing the desired constraints (but note this can be tricky when attempting to use both `workBounds` and `userBounds` at the same time). See #57 and #180, as well as the many examples in the [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf).
